### PR TITLE
feat(x/ecocredit): sdk.Msg for MsgTake

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Ref: #732

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/basket/msg_take.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	if _, err := sdk.AccAddressFromBech32(m.Owner); err != nil {
+		return sdkerrors.ErrInvalidRequest.Wrapf(err.Error())
+	}
+	if err := sdk.ValidateDenom(m.BasketDenom); err != nil {
+		return sdkerrors.ErrInvalidRequest.Wrapf("%s is not a valid basket denom", m.BasketDenom)
+	}
+	if _, err := math.NewPositiveDecFromString(m.Amount); err != nil {
+		return sdkerrors.ErrInvalidRequest.Wrap(err.Error())
+	}
+	if err := ecocredit.ValidateLocation(m.RetirementLocation); err != nil {
```
Comment (technicallyty) [inline x/ecocredit/basket/msg_take.go:39]: we should check if its empty first, as its an optional field

File: x/ecocredit/basket/msg_take_test.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+			fields: fields{
+				Owner:              addr.String(),
+				BasketDenom:        "BCT",
+				Amount:             "-12.34",
+				RetirementLocation: "US-WA",
+			},
+			wantErr: true,
+		},
+		{
+			name: "invalid retirement location - empty",
```
Comment (technicallyty) [inline x/ecocredit/basket/msg_take_test.go:113]: this should be a valid case

**Review by technicallyty — APPROVED** [review]
looks good! just an issue with location

File: x/ecocredit/basket/msg_take.go (hunk 1 above)
Comment (ryanchristo) [reply x/ecocredit/basket/msg_take.go]: Updated to validate if `RetireOnTake` is true.

**Review by clevinson — APPROVED** [review]

File: x/ecocredit/basket/msg_take_test.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+				BasketDenom:        "BCT",
+				Amount:             "12.34",
+				RetirementLocation: "",
+				RetireOnTake:       true,
+			},
+			wantErr: true,
+		},
+		{
+			name: "invalid retirement location - format",
+			fields: fields{
```
Comment (robert-zaremba) [inline x/ecocredit/basket/msg_take_test.go:145]: You could use `MsgTake` directly here instead of introducing `fields`.

**Pull request merged**